// variant/lib/slim.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.payload = exports.nil = exports.fields = exports.HOI = exports.pass = exports.patterned = exports.unpack = exports.just = exports.constant = exports.literalist = exports.catalog = exports.typeCatalog = exports.typeMap = exports.onTerms = exports.variantCosmos = exports.construct = exports.constrained = exports.augment = void 0;
var augment_1 = require("./augment");
Object.defineProperty(exports, "augment", { enumerable: true, get: function () { return augment_1.augment; } });
var constrained_1 = require("./constrained");
Object.defineProperty(exports, "constrained", { enumerable: true, get: function () { return constrained_1.constrained; } });
var construct_1 = require("./construct");
Object.defineProperty(exports, "construct", { enumerable: true, get: function () { return construct_1.construct; } });
var cosmos_1 = require("./cosmos");
Object.defineProperty(exports, "variantCosmos", { enumerable: true, get: function () { return cosmos_1.variantCosmos; } });
var generic_1 = require("./generic");
Object.defineProperty(exports, "onTerms", { enumerable: true, get: function () { return generic_1.onTerms; } });
var typeCatalog_1 = require("./typeCatalog");
Object.defineProperty(exports, "typeMap", { enumerable: true, get: function () { return typeCatalog_1.typeMap; } });
Object.defineProperty(exports, "typeCatalog", { enumerable: true, get: function () { return typeCatalog_1.typeCatalog; } });
var catalog_1 = require("./catalog");
Object.defineProperty(exports, "catalog", { enumerable: true, get: function () { return catalog_1.catalog; } });
Object.defineProperty(exports, "literalist", { enumerable: true, get: function () { return catalog_1.literalist; } });
var match_tools_1 = require("./match.tools");
Object.defineProperty(exports, "constant", { enumerable: true, get: function () { return match_tools_1.constant; } });
Object.defineProperty(exports, "just", { enumerable: true, get: function () { return match_tools_1.just; } });
Object.defineProperty(exports, "unpack", { enumerable: true, get: function () { return match_tools_1.unpack; } });
var patterned_1 = require("./patterned");
Object.defineProperty(exports, "patterned", { enumerable: true, get: function () { return patterned_1.patterned; } });
var typed_1 = require("./typed");
Object.defineProperty(exports, "pass", { enumerable: true, get: function () { return typed_1.pass; } });
var util_1 = require("./util");
Object.defineProperty(exports, "HOI", { enumerable: true, get: function () { return util_1.HOI; } });
var variant_tools_1 = require("./variant.tools");
Object.defineProperty(exports, "fields", { enumerable: true, get: function () { return variant_tools_1.fields; } });
Object.defineProperty(exports, "nil", { enumerable: true, get: function () { return variant_tools_1.nil; } });
Object.defineProperty(exports, "payload", { enumerable: true, get: function () { return variant_tools_1.payload; } });
